cooked-validators/src/Cooked/Tx/Constraints/Type.hs

```diff
@@ -107,13 +120,34 @@ data MiscConstraint where
   -- | Ensure that the transaction is signed by the given 'Pl.PubKeyHash'es.
   SignedBy :: [Pl.PubKeyHash] -> MiscConstraint
 
+-- NB don't forget to update the Eq instance when adding new constructors
+
+(~*~?) :: forall ty a1 a2. (Typeable a1, Typeable a2) => ty a1 -> ty a2 -> Maybe (a1 :~~: a2)
+_ ~*~? _ = typeRep @a1 `eqTypeRep` typeRep @a2
```

Comment:
That's clever! (and solves the problem where I got stuck and decided to take the  `toBuiltinData` route)